### PR TITLE
Port NGHTTP2 on TizenRT platform

### PR DESCRIPTION
**Pull request by sahilpai1**
@juitem , @sunghan-chang, @pillip8282, @bhargava-cs
Please review NGHTTP2 Patch

All review comments from the below pull request have been addressed:
https://github.com/Samsung/TizenRT/pull/1479

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [255dcf0107341234b9cff9269f27bba8e3dbe796] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [255dcf0107341234b9cff9269f27bba8e3dbe796] - Code Rule Check (C++) OK.

File: external/nghttp2/nghttp2_debug.c [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "nghttp2_debug.h"
+
+#include <stdio.h>
+
+#ifdef CONFIG_NGHTTP2_DEBUG
```
Comment (sunghan-chang) [inline external/nghttp2/nghttp2_debug.c]: To check CONFIG_XX definition, you MUST include <tinyara/config.h>

File: external/nghttp2/Kconfig [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+	bool "Enable NGHTTP2"
+	default n
+	depends on MBEDTLS
+	---help---
+		nghttp2 is an implementation of HTTP/2 and its header compression algorithm HPACK.
+
+if ENABLE_NGHTTP2
+config NGHTTP2_DEBUG
+	bool "Nghttp2 Debug"
+	default n
```
Comment (sunghan-chang) [inline external/nghttp2/Kconfig]: How about make a dependancy with DEBUG?
If it has, when we disable DEBUG, all of debug feature will be disabled automatically. User does not need to check all of debug feature at each module.
```
depends on DEBUG
```

File: external/nghttp2/nghttp2_debug.c (hunk 1 above)
Comment (sahilpai1) [reply external/nghttp2/nghttp2_debug.c]: @sunghan-chang ok. Adding #include <tinyara/config.h> to 'nghttp2_debug.h'

File: external/nghttp2/Kconfig (hunk 2 above)
Comment (sahilpai1) [reply external/nghttp2/Kconfig]: @sunghan-chang  ok, made changes.
But with 'default n' i dont think this issue would arise

Comment (sunghan-chang) [reply external/nghttp2/Kconfig]: That is a default value. When config does not have any value, default value will be set. But I concern the case that nghttp2 debug is set to y. If we keep it as of now, we should turn off debug and nghttp2 debug separately. But if we make a dependancy, just turnning off debug is enough to disable all of debug feature.

Comment (sahilpai1) [reply external/nghttp2/Kconfig]: @sunghan-chang Yes, true. Made changes

**Comment by seinfra** [comment]
Target : [7c4ac0355ee15a59c10233d90cb9be79192cb29c] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [7c4ac0355ee15a59c10233d90cb9be79192cb29c] - Code Rule Check OK.

**Comment by SeonghoByeon** [comment]
Currently, we cannot test the patch properly due to the curl build error. 
Please elaborates on how to check the patch.

**Comment by sahilpai1** [comment]
@SeonghoByeon I am not getting any build breaks. Could you please share the error logs.

**Comment by SeonghoByeon** [comment]
@sahilpai1 Now I have successfully built this patch without curl problem. 
One of the merged commits was missing (sorry for the confusion.).
BTW, when I configure the nghttp2-enabled build with grpc, I could see the following error.

...
AR:     helloxx_main.o
make[2]: Leaving directory '/home/samsung/shbyeon/TizenRT/apps/examples/helloxx'
make[2]: Entering directory '/home/samsung/shbyeon/TizenRT/apps/examples/grpc_greeter_client'
protoc -I . --cpp_out=. helloworld.proto
make[2]: protoc: Command not found
Makefile:86: recipe for target 'helloworld.pb.cc' failed
make[2]: *** [helloworld.pb.cc] Error 127
make[2]: Leaving directory '/home/samsung/shbyeon/TizenRT/apps/examples/grpc_greeter_client'
Makefile:109: recipe for target 'examples/grpc_greeter_client_all' failed
make[1]: *** [examples/grpc_greeter_client_all] Error 2
make[1]: Leaving directory '/home/samsung/shbyeon/TizenRT/apps'
LibTargets.mk:158: recipe for target '../apps/libapps.a' failed
make: *** [../apps/libapps.a] Error 2

**Comment by pillip8282** [comment]
@sahilpai1 
Could you share how to run nghttp?. We are supposed to verify it before merge it.

**Comment by seinfra** [comment]
Target : [1634fcc81a4f534b593cad25e3834607d583d3f4] - Code Rule Check OK.

**Comment by SeonghoByeon** [comment]
@sahilpai1 @pillip8282 I run nghttp2 using curl app as below, following your guide.
Please check the log. (Failed HTTPS connection due to mbedTLS certification.)

TASH>>ifconfig wl1 dhcp
IP address 192.168.1.2
Netmask 255.255.255.0
Gateway 192.168.1.1
TASH>>date -s APR 27 10:00:00 2018 Fri 104 0
TASH>>curl http2 get https://www.google.com
Command will be launched with pri (100), stack size(4096)
TASH>>Curl INFO STATE: INIT => CONNECT handle 0x2078ab0; line 1392 (connection #-5000)
Curl INFO Marked for [closure]: Default to force-close
Curl INFO Rebuilt URL to: https://www.google.com/
Curl INFO Added connection 0. The cache now contains 1 members
Curl INFO STATE: CONNECT => WAITRESOLVE handle 0x2078ab0; line 1428 (connection #0)
Curl INFO   Trying 74.125.203.104...
Curl INFO TCP_NODELAY set
Curl INFO STATE: WAITRESOLVE => WAITCONNECT handle 0x2078ab0; line 1509 (connection #0)
Curl INFO Curl_poll(1 ds, 200 ms) == 1
Curl INFO STATE: WAITCONNECT => SENDPROTOCONNECT handle 0x2078ab0; line 1561 (connection #0)
Curl INFO Marked for [keep alive]: HTTP default
Curl FAIL Error reading ca cert file /rom/curl-certificates.crt - mbedTLS: (-0x3E00) PK - Read/write of file failedCurl INFO Marked for [closure]: Failed HTTPS connection
Curl INFO STATE: SENDPROTOCONNECT => PROTOCONNECT handle 0x2078ab0; line 1575 (connection #0)
Curl INFO Expire cleared
Curl INFO multi_done
Curl INFO Closing connection 0
Curl INFO The cache now contains 0 members
ret [77]
curl test is completed

**Comment by sahilpai1** [comment]
Hi @SeonghoByeon,
Currently the design is, 'curl-certificates.crt' is present in 'external/curl' directory.
When curl is supported, then after compiling code this 'curl-certificates.crt' is moved into 'tools/fs/contents' directory.
Now when it is flashed onto board, this .crt file is moved into 'rom' directory of board.

It is upto the board to support rom fs, without which an error will be thrown like in your case.
In T200 board we are not having any issue.

**Comment by seinfra** [comment]
Target : [b17c4d28b288c8b07c9ad3df8e9ba51a1986874e] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [b17c4d28b288c8b07c9ad3df8e9ba51a1986874e] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [1634fcc81a4f534b593cad25e3834607d583d3f4] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [1634fcc81a4f534b593cad25e3834607d583d3f4] - Code Rule Check OK.

**Comment by SeonghoByeon** [comment]
Hi @sahilpai1,
thanks for the guidance for certi file. 
Artik053 board supports romfs, which was turned off in the previous test.
Now, I double-check that crt file exists in rom directory on board. 

First trial (google) seems to be successfully finished,
but second trial (google) fails HTTP connection,
and the last one (facebook) finally causes an assertion failure.
Please see below.

TASH>>curl http2 get https://www.google.com
Command will be launched with pri (100), stack size(4096)
TASH>>Curl INFO STATE: INIT => CONNECT handle 0x207a490; line 1392 (connection #-5000)
Curl INFO Marked for [closure]: Default to force-close
Curl INFO Rebuilt URL to: https://www.google.com/
Curl INFO Added connection 0. The cache now contains 1 members
Curl INFO STATE: CONNECT => WAITRESOLVE handle 0x207a490; line 1428 (connection #0)
Curl INFO   Trying 74.125.203.147...
Curl INFO TCP_NODELAY set
Curl INFO STATE: WAITRESOLVE => WAITCONNECT handle 0x207a490; line 1509 (connection #0)
Curl INFO Curl_poll(1 ds, 200 ms) == 1
Curl INFO STATE: WAITCONNECT => SENDPROTOCONNECT handle 0x207a490; line 1561 (connection #0)
Curl INFO Marked for [keep alive]: HTTP default
Curl INFO mbedTLS: Connecting to www.google.com:443
Curl INFO mbedTLS: Set min SSL version to TLS 1.0
Curl INFO ALPN, offering h2
Curl INFO ALPN, offering http/1.1
Curl INFO STATE: SENDPROTOCONNECT => PROTOCONNECT handle 0x207a490; line 1575 (connection #0)
Curl INFO Curl_poll(1 ds, 120 ms) == 1
Curl INFO Curl_poll(1 ds, 50 ms) == 1
Curl INFO Curl_poll(1 ds, 0 ms) == 0
curl http2 get https://www.google.com
Command will be launched with pri (100), stack size(4096)
TASH>>Curl INFO STATE: INIT => CONNECT handle 0x209d6a0; line 1392 (connection #-5000)
Curl INFO Marked for [closure]: Default to force-close
Curl INFO Rebuilt URL to: https://www.google.com/
Curl INFO Added connection 0. The cache now contains 1 members
Curl INFO   Trying 74.125.203.147...
Curl INFO TCP_NODELAY set
Curl INFO STATE: CONNECT => WAITCONNECT handle 0x209d6a0; line 1444 (connection #0)
Curl INFO Curl_poll(1 ds, 0 ms) == 0
Curl INFO Curl_poll(1 ds, 200 ms) == 1
Curl INFO STATE: WAITCONNECT => SENDPROTOCONNECT handle 0x209d6a0; line 1561 (connection #0)
Curl INFO Marked for [keep alive]: HTTP default
Curl INFO mbedTLS: Connecting to www.google.com:443
Curl INFO mbedTLS: Set min SSL version to TLS 1.0
Curl INFO ALPN, offering h2
Curl INFO ALPN, offering http/1.1
Curl INFO STATE: SENDPROTOCONNECT => PROTOCONNECT handle 0x209d6a0; line 1575 (connection #0)
Curl INFO Curl_poll(1 ds, 90 ms) == 1
Curl INFO Curl_poll(1 ds, 1000 ms) == 1
Curl INFO mbedTLS: Handshake complete, cipher is TLS-ECDHE-ECDSA-WITH-AES-128-GCM-SHA256
Curl FAIL Cert verify failed: BADCERT_NOT_TRUSTEDCurl INFO Marked for [closure]: Failed HTTPS connection
Curl INFO multi_done
Curl INFO stopped the pause stream!
Curl INFO Closing connection 0
Curl INFO The cache now contains 0 members
ret [51]
curl test is completed
curl http2 get https://www.facebook.com
Command will be launched with pri (100), stack size(4096)
TASH>>Curl INFO STATE: INIT => CONNECT handle 0x209d6a0; line 1392 (connection #-5000)
Curl INFO Marked for [closure]: Default to force-close
Curl INFO Rebuilt URL to: https://www.facebook.com/
Curl INFO Added connection 0. The cache now contains 1 members
arm_dataabort: 
Data abort. PC: 040d4dd0 DFAR: 363f1e44 DFSR: 00000008
up_assert: Assertion failed at file:armv7-r/arm_dataabort.c line: 111 task: <pthread>
up_dumpstate: Current sp: 0209d458
up_dumpstate: User stack:
up_dumpstate:   base: 0209d688
up_dumpstate:   size: 000027fc
up_dumpstate:   used: 0000132c
up_dumpstate: User Stack
up_stackdump: 0209d440: 0209de30 040c988c 0209de30 040c988c 0209de30 040c988c 0209de30 040c988c
up_stackdump: 0209d460: 363f1e44 00000008 00000008 040c827c ffffffff 0202b7a0 cbd20330 363f1e40
up_stackdump: 0209d480: 34340020 020b1e30 0202b7a0 020b1e20 0209de30 00000000 04207a53 00000000
up_stackdump: 0209d4a0: ffffffff 00000000 0209d4b8 040d4dbc 040d4dd0 a0000053 02024920 0209d6a0
up_stackdump: 0209d4c0: 020af2a0 040fa6fc 000493e0 020b1bf0 00000000 0209d538 03c0a7c7 041d2f68
up_stackdump: 0209d4e0: 00000000 020b1e70 0209d6a8 0209d5b2 00000000 020a16a0 0209d620 00000000
up_stackdump: 0209d500: 020af480 041d2e58 020af4a4 0209d5b3 000e9c47 020af53c 020af54c 00000002
up_stackdump: 0209d520: 0000003e 000e9c47 00000001 020b1e30 020b1e50 00000000 00000000 00000000
up_stackdump: 0209d540: 0000003e 000e9c47 00000000 00000000 0000003e 39127558 041e60b8 0209d6a0
up_stackdump: 0209d560: ffffffff 00000000 020a62e0 020a16a0 0209d620 00000000 0209d628 04101d68
up_stackdump: 0209d580: 0209d628 041d2e58 0209d590 041d4c00 000e9c47 00000000 0209d5f0 0209d5f0
up_stackdump: 0209d5a0: 00000000 041d4974 0000003e 000e9c47 000103e7 02090000 0209d5f0 00000000
up_stackdump: 0209d5c0: 00000001 020a62e0 0209d5f0 00000000 0209d6a0 0209d618 0209d620 0209d61c
up_stackdump: 0209d5e0: 0209d628 04102c28 020a62e0 0209d618 0000003e 000e9c47 020a62e0 0209d6a0
up_stackdump: 0209d600: 00000000 00000001 00000001 040f1314 0209d61c 040f7890 00000000 00000000
up_stackdump: 0209d620: 0000003e 000e9c47 0000003e 000e9c47 ffffffbf 0209acfc 0209d6a0 ffffffbf
up_stackdump: 0209d640: 0209986c 00000000 00000000 00000000 00000000 040dd060 0206a250 02099860
up_stackdump: 0209d660: 0209ad50 0209986c 00000000 041d3ccc 00000000 00000000 00000000 00000000
up_stackdump: 0209d680: 00000000 00000000 deadbeef 00000000 00004b10 80002810 040f9194 00000012
up_registerdump: R0: 0202b7a0 cbd20330 363f1e40 34340020 020b1e30 0202b7a0 020b1e20 0209de30
up_registerdump: R8: 00000000 04207a53 00000000 ffffffff 00000000 0209d4b8 040d4dbc 040d4dd0
up_registerdump: CPSR: a0000053
up_showtasks: *******************************************
up_showtasks: List of all tasks in the system:
up_showtasks: *******************************************
up_taskdump: Idle Task: PID=0 Stack Used=0 of 0
up_taskdump: hpwork: PID=1 Stack Used=636 of 2028
up_taskdump: lpwork: PID=2 Stack Used=468 of 2028
up_taskdump: LWIP_TCP/IP: PID=3 Stack Used=892 of 4068
up_taskdump: tash: PID=5 Stack Used=1652 of 4076
up_taskdump: WPA Supplicant: PID=7 Stack Used=1740 of 4028
up_taskdump: WLAN Driver mxmgmt: PID=8 Stack Used=244 of 2044
up_taskdump: WPA Ctrl Iface FIFO: PID=9 Stack Used=428 of 2044
up_taskdump: Wi-Fi API monitor: PID=10 Stack Used=756 of 2044
up_taskdump: curl: PID=11 Stack Used=228 of 4028
up_taskdump: <pthread>: PID=12 Stack Used=5124 of 10236
up_taskdump: curl: PID=17 Stack Used=228 of 4028
up_taskdump: <pthread>: PID=18 Stack Used=4908 of 10236
up_taskdump: <pthread>: PID=19 Stack Used=388 of 2044

**Comment by sahilpai1** [comment]
Hi @SeonghoByeon @pillip8282 ,

_but second trial (google) fails HTTP connection,_
As mentioned earlier, if you get any error during TLS handshake it is because the system date has not been updated.
Update date using following command:
**date -s APR 27 10:00:00 2018 Fri 104 0**
Also, each time the board is reset, the date goes back to 1970. So the date needs to be updated every time you reset.

_and the last one (facebook) finally causes an assertion failure._
Ok this is actually happening due to low stack size provided to curl's pthread.
I have raised a patch https://github.com/Samsung/TizenRT/pull/1626 now.
Once it gets merged i will let you know. I am positive even this test case will work fine.

**Comment by sahilpai1** [comment]
Hi @pillip8282,

Yes there was a config.h file with nghttp2.
But i noticed none of the macros were being used anywhere in the rest of the code.
So i thought it is not needed. Could you pls tell me what is the requirement of it. 

I am adding a new commit to this PR for now, having the config file.
Also i have enabled HAVING_CONFIG_H macro in Makefile so that it is included in .h files

**Comment by seinfra** [comment]
Target : [32dd5ba26fa089411efc5ddd14d86b265bcf4998] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [32dd5ba26fa089411efc5ddd14d86b265bcf4998] - Code Rule Check OK.

**Comment by sahilpai1** [comment]
Hi @SeonghoByeon, @pillip8282 ,

_and the last one (facebook) finally causes an assertion failure_
Ok the patch for increasing stack size has been merged - #1626 
Could you pls verify with facebook again. I am sure it will work now. Thanks

**Comment by SeonghoByeon** [comment]
Hi @sahilpai1, @pillip8282,
Please find the attached file below,
which is obtained by running curl app targeted to facebook using http2.
Is it an expected result?
(Certain phrases are repeated endlessly.)
[cutecom.log](https://github.com/Samsung/TizenRT/files/1974032/cutecom.log)

Moreover, 
"curl http2 get https://www.google.com" does not work properly.
Please see the following log.
(crt file exists in /rom/ directory.)

TASH>>date -s APR 27 10:00:00 2018 Fri 104 0
TASH>>curl http2 get https://www.google.com
Command will be launched with pri (100), stack size(4096)
TASH>>Curl INFO STATE: INIT => CONNECT handle 0x209b510; line 1392 (connection #-5000)
Curl INFO Marked for [closure]: Default to force-close
Curl INFO Rebuilt URL to: https://www.google.com/
Curl INFO Added connection 0. The cache now contains 1 members
Curl INFO   Trying 74.125.203.147...
Curl INFO TCP_NODELAY set
Curl INFO STATE: CONNECT => WAITCONNECT handle 0x209b510; line 1444 (connection #0)
Curl INFO Curl_poll(1 ds, 0 ms) == 0
Curl INFO Curl_poll(1 ds, 200 ms) == 1
Curl INFO STATE: WAITCONNECT => SENDPROTOCONNECT handle 0x209b510; line 1561 (connection #0)
Curl INFO Marked for [keep alive]: HTTP default
Curl INFO mbedTLS: Connecting to www.google.com:443
Curl INFO mbedTLS: Set min SSL version to TLS 1.0
Curl INFO ALPN, offering h2
Curl INFO ALPN, offering http/1.1
Curl INFO STATE: SENDPROTOCONNECT => PROTOCONNECT handle 0x209b510; line 1575 (connection #0)
Curl INFO Curl_poll(1 ds, 80 ms) == 1
Curl INFO Curl_poll(1 ds, 1000 ms) == 1
Curl INFO mbedTLS: Handshake complete, cipher is TLS-ECDHE-ECDSA-WITH-AES-128-GCM-SHA256
Curl FAIL Cert verify failed: BADCERT_NOT_TRUSTEDCurl INFO Marked for [closure]: Failed HTTPS connection
Curl INFO multi_done
Curl INFO stopped the pause stream!
Curl INFO Closing connection 0
Curl INFO The cache now contains 0 members
ret [51]
curl test is completed

**Comment by sahilpai1** [comment]
@SeonghoByeon 
Ok i also just saw ur previous comment, where you mentioned "curl http2 get https://www.google.com" has worked.
No, it hasn't completed for you. The logs you should get is [http2-google.log](https://github.com/Samsung/TizenRT/files/1974266/http2-google.log)

Also, I have enabled debug logs for nghttp2 in this...

**Comment by SeonghoByeon** [comment]
Hi @sahilpai1 @pillip8282,
Finally, every command works properly.
Thank you for your kind explanation with patience.

**Comment by sahilpai1** [comment]
Hi @SeonghoByeon @pillip8282 
That's great. Thanks

**Comment by pillip8282** [comment]
@sahilpai1 
I think using config.h make nghttp2 portable. for example, if TizenRT doesn't support dup2(it is not supported now as lwip 2.0 upgrade) then we just remove HAVE_DUP2 configs. 
and dup2 is not supported now on TizenRT so plz remove that in config.h

**Comment by seinfra** [comment]
Target : [6a27437c9f3f08de1e443ba4f7ee26e464f23bce] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [6a27437c9f3f08de1e443ba4f7ee26e464f23bce] - Code Rule Check OK.

File: external/nghttp2/nghttp2_config.h [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+/* Define to 1 if you have the declaration of `strerror_r', and to 0 if you
+   don't. */
+#define HAVE_DECL_STRERROR_R 1
+
+/* Define to 1 if you have the <dlfcn.h> header file. */
+#define HAVE_DLFCN_H 1
+
+/* Define to 1 if you have the `dup2' function. */
+#ifndef __TIZENRT__
+#define HAVE_DUP2 1
```
Comment (pillip8282) [inline external/nghttp2/nghttp2_config.h]: I think just commenting HAVE_DUP2 is enough without __TIZENRT__
because every platform will have their own.

Comment (sahilpai1) [reply external/nghttp2/nghttp2_config.h]: every platform will have their own, meaning other platforms wudn't use HAVE_DUP2 at all?

Comment (pillip8282) [reply external/nghttp2/nghttp2_config.h]: It would be better to modify it as below
/* #define HAVE_DUP2 1 */
because config.h is an auto-generated file

Comment (sahilpai1) [reply external/nghttp2/nghttp2_config.h]: @pillip8282 Ok. done

**Comment by seinfra** [comment]
Target : [52a55818bae97946f4da547a6c997b77c3b3a335] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [52a55818bae97946f4da547a6c997b77c3b3a335] - Code Rule Check OK.

**Comment by sahilpai1** [comment]
@pillip8282 ok

**Comment by seinfra** [comment]
Target : [52a55818bae97946f4da547a6c997b77c3b3a335] -

**Pull request merged**